semantics/src/traversal/functions.rs

```diff
@@ -94,19 +94,34 @@ pub fn func_def(
 }
 
 fn validate_all_paths_return_or_revert(
-    body: &[Spanned<fe::FuncStmt>],
+    block: &[Spanned<fe::FuncStmt>],
 ) -> Result<(), SemanticError> {
-    // This will need to become more sophisticated when we introduce branching logic
-    // because we then need to follow different code paths and check that they
-    // all return or revert.
-    for statement in body {
+    for statement in block {
         if let fe::FuncStmt::Return { .. } = &statement.node {
             return Ok(());
         }
 
         if let fe::FuncStmt::Revert { .. } = &statement.node {
             return Ok(());
         }
+
+        if let fe::FuncStmt::If {
+            test: _,
+            body,
+            or_else,
+        } = &statement.node
+        {
+            let body_returns = validate_all_paths_return_or_revert(body)?;
+            let or_else_returns = if !or_else.is_empty() {
+                validate_all_paths_return_or_revert(or_else)?;
+            };
+            if vec![body_returns, or_else_returns]
+                .iter()
+                .all(|result| result == &())
```

Comment:
Not, super proud of this. There's probably a more rusty way to do this. I thought about something like:

```rust
[
validate_all_paths_return_or_revert(body),
validate_all_paths_return_or_revert(or_else),
].collect::<Something, Something>()?;
```
But I gave up on that since the `or_else` needs to be checked for emptiness.

Reply:
If I understand correctly, the following would fail to compile:

```
contract Foo:
    pub def bar(val: u256) -> u256:
        if val > 1:
            return 5
        else:
            x = 1
        
        return 42
```

as ` validate_all_paths_return_or_revert(or_else)?` would return an error before we get to the final statement in `bar`. This could probably be fixed and cleaned up by doing something like this:

```rust
let body_returns = validate_all_paths_return_or_revert(body).is_ok();
let or_else_returns = if !or_else.is_empty() {
  validate_all_paths_return_or_revert(or_else).is_ok();
};

if body_returns && or_else_return {
  // code definitely terminates
  return Ok(());
}
// continue checking statements
```

However, I think calling `validate_all_paths_return_or_revert` on code blocks that don't need to terminate is a bit misleading. It might be better to have separate function named `is_terminal` that takes a list of statements and returns a bool, instead of a result. The base case of this function would be `false`, and any absolutely terminal statements that we encounter would return `true`. We could then wrap this function inside of `validate_all_paths_return_or_revert` where it expects a terminal code block and fails if it does not get one. Does this seem cleaner?

Reply:
Something worth noting is that `if` statements without an `or_else` are not guaranteed to terminate, so we don't need to check the `if` code block in that case. We only need to check the contents of an `if` statement if there is an `or_else` attached. In that case, we would check to make sure that both blocks are guaranteed to terminate.

Even in the case of:

```
def bar() -> u256:
  if true:
    return 42
```

We would throw an error since sematic analysis isn't concerned about what values expressions evaluate to.

With something like this:

```
def bar(x: u256) -> u256:
  # non-terminal
  if <case1>:
    return 42

  # terminal
  if <case2>:
    # terminal
    return 26
  else:
    # non-terminal
    if <case3>:
      return 4226

    # terminal
    return 9999

```

We can just think about whether or not a statement is terminal or non-terminal and require that all code branches end with a terminal statement.

Reply:
Ah, you are making great points here. The snippet you shared is in fact falsely categorized as *not returning* even though it clearly does.

>is there any reason to iterate over each statement, or can we just recursively grab the last one and make sure it terminates?

That's an interesting point. I think that would certainly make things easier to implement. That said, in Rust (as well as many other languages) such code would be valid and just trigger a warning because of an unreachable statement.

```rust
fn something() -> bool {
    return true;
    let x = 1;
}
```

In our case, it would lead to a hard error because the last statement would not be a `return` or `revert`.

I think I have to play a bit with the code to get to an informed opinion :)

Reply:
Regarding my previous comment. If we don't want to fixate on the last statement of the block (which would diverge from what other languages are doing), I think the closest do that would be to just check the statements in reverse order and return `Ok(())` as soon as we find the first termination. I added a test case to cover the case that previously failed.